### PR TITLE
feat: introduce max change per trade

### LINKED ISSUE #226: Caveats regarding the clamp
Say on an L2 (e.g. Arbitrum), if the sequencer goes down for a few hours, no swap would be able to take place. As such, the first swap that happens can swing the instant clamped (and of course raw) price a lot. In other words, the clamp becomes quite useless if there are no swaps for an extended period of time, or put it inversely, the clamp works best when there are frequent swaps.

e.g. with a 1bp/s rate of change, if there are no swaps for 2 hours (7200 s), the swap can move the price by 72% and it would not exceed the clamp.

Potential mitigations: 
1. New borrows use the instant price in the case where there has been no oracle observation for X period 
2. Liquidations use a longer TWAP (e.g. 1-hr TWAP) instead of the normal TWAP (10-15 mins)
3. Clamp the max change within one trade
  - 5% ?

### PR DESCRIPTION
**Pull request by xenide**
## Solution

with the implementation of this extra check (max change per trade), there are 4 possible scenarios: 

1. raw price doesn't violate both conditions
  - simple: clampedPrice = rawPrice 
2. raw price only violates rate of change, doesn't violate max change per trade
  - use the rate-of-change capped price  
  - is it possible that the max-change-per-trade capped price is a smaller change than rate-of-change-capped price? 
    - e.g. maxChangePerTrade is 2%, rate of change is 3bp/s. 
      - trade comes after 30 seconds, doing a 1% change in price. 
      - max-change-per-trade capped price is 102%, rate-of-change capped price is 100.9%
    - e.g. maxChangePerTrade is 2%, rate of change is 100bp/s. 
      - trade comes after 3 seconds, doing a 1% change in price
      - max-change-per-trade capped price is 102%, rate-of-change capped price is 103%
    - seems that it's not possible
3. raw price only violates max change per trade, doesn't violate rate of change
  - use the max-change-per-trade capped price 
4. raw violates both rate of change and max change per trade
  - should we take the rate-of-change capped or max-change-per-trade capped? 
  - should choose the closer one? Arithmetic / geometric mean of the two?
    - Arithmetic mean would be cheaper than geometric mean due to gas cost 
  - Propose to choose the closer one (smaller change)

### CODE REVIEW

File: src/ReservoirPair.sol [hunk 1]
```diff
@@ -557,19 +565,37 @@ abstract contract ReservoirPair is IAssetManagedPair, ReservoirERC20 {
...
             // aTimeElapsed <= 32 bits
             if (aCurrRawPrice > aPrevClampedPrice) {
-                rClampedPrice = aPrevClampedPrice.fullMulDiv(1e18 + maxChangeRate * aTimeElapsed, 1e18);
+                uint256 lRateLimitedPrice = aPrevClampedPrice.fullMulDiv(1e18 + maxChangeRate * aTimeElapsed, 1e18);
+                uint256 lPerTradeLimitedPrice = aPrevClampedPrice.fullMulDiv(1e18 + maxChangePerTrade, 1e18);
+                rClampedPrice = lRateLimitedPrice.min(lPerTradeLimitedPrice);
+                assert(rClampedPrice < aCurrRawPrice);
             } else {
-                assert(aPrevClampedPrice > aCurrRawPrice);
-                rClampedPrice = aPrevClampedPrice.fullMulDiv(1e18 - maxChangeRate * aTimeElapsed, 1e18);
```
Comment (xenide) [inline src/ReservoirPair.sol:574]: I think the auditors missed this. This underflow vuln has got to be a critical, and is live on our production pairs right now

File: src/ReservoirPair.sol [hunk 2]
```diff
@@ -557,19 +565,37 @@ abstract contract ReservoirPair is IAssetManagedPair, ReservoirERC20 {
...
             (rClampedPrice, rClampedLogPrice) = (aCurrRawPrice, aCurrLogRawPrice);
         } else {
             // clamp the price
             // multiplication of maxChangeRate and aTimeElapsed would not overflow as
             // maxChangeRate <= 0.01e18 (50 bits)
             // aTimeElapsed <= 32 bits
             if (aCurrRawPrice > aPrevClampedPrice) {
-                rClampedPrice = aPrevClampedPrice.fullMulDiv(1e18 + maxChangeRate * aTimeElapsed, 1e18);
+                uint256 lRateLimitedPrice = aPrevClampedPrice.fullMulDiv(1e18 + maxChangeRate * aTimeElapsed, 1e18);
+                uint256 lPerTradeLimitedPrice = aPrevClampedPrice.fullMulDiv(1e18 + maxChangePerTrade, 1e18);
```
Comment (xenide) [inline src/ReservoirPair.sol]: I think an optimization we can do here, is to first calculate the lower rate of change, and then do only one `fullMulDiv`, as a `fullMulDiv` is expensive

Comment (OliverNChalk) [reply src/ReservoirPair.sol]: You don't need to clamp the output, you can just clamp the multiplier. I.e. `min(maxChangeRate * aTimeElapsed, maxChangePerTrade)`

File: src/ReservoirPair.sol (hunk 1 above)
Comment (OliverNChalk) [reply src/ReservoirPair.sol:574]: What's the vuln, it freezes the pair if it overflows? How realistic is 1e18 seconds elapsed, surely that's like super far away?

Comment (xenide) [reply src/ReservoirPair.sol:574]: it's the product of `maxChangeRate` and `aTimeElapsed` that's easy to overflow `1e18`. 

e.g. 86400 (seconds in a day) * 0.00003e18 (0.3bp / s)  = 2.592e18 > 1e18

Yes it freezes the pair as it's an arithmetic underflow. I discovered it during regression testing. Kinda weird that it didn't show up before

File: src/ReservoirPair.sol (hunk 2 above)
Comment (xenide) [reply src/ReservoirPair.sol]: yup totally. Implemented

**Pull request merged**